### PR TITLE
RSDK-2262 - use config to determine slam mode during validation

### PR DESCRIPTION
**Pull request by kim-mishra**
### Changes
* throw an error if config suggests `localization mode` & `offline mode`
* update tests to test above change

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -161,8 +165,13 @@ func GetOptionalParameters(config *Config, defaultLidarDataRateMsec, defaultIMUD
 		if config.EnableMapping == nil {
 			logger.Debug("no enable_mapping given, setting to default value of false")
 		} else {
-			optionalConfigParams.EnableMapping = config.CloudStoryEnabled
+			optionalConfigParams.EnableMapping = *config.EnableMapping
```
Comment (kim-mishra) [inline config/config.go:168]: this was a bug

File: config/config.go [hunk 2]
```diff
@@ -179,3 +188,18 @@ func GetOptionalParameters(config *Config, defaultLidarDataRateMsec, defaultIMUD
 
 	return optionalConfigParams, nil
 }
+
+func validateModes(
+	optionalConfigParams OptionalConfigParams,
+	imuEnabled bool,
+) error {
+	localizationMode := optionalConfigParams.LidarDataRateMsec == 0
```
Comment (EshaMaharishi) [inline config/config.go:196]: This seems wrong - isn't it that if LidarDataRateMsec == 0, we're in *offline* mode, not *localization* mode?

File: config/config.go [hunk 3]
```diff
@@ -179,3 +188,18 @@ func GetOptionalParameters(config *Config, defaultLidarDataRateMsec, defaultIMUD
...
+	offlineMode := !optionalConfigParams.EnableMapping && optionalConfigParams.ExistingMap != ""
```
Comment (EshaMaharishi) [inline config/config.go:197]: Similarly here, should this be setting localizationMode?

File: config/config.go (hunk 2 above)
Comment (kim-mishra) [reply config/config.go:196]: yes got them backwards fixing!

File: config/config.go (hunk 3 above)
Comment (kim-mishra) [reply config/config.go:197]: yes got them backwards fixing!

[the remark this answers is not in the document]
Comment (kim-mishra) [reply config/config.go:197]: oh you are right - removing `existing_map` check

File: config/config.go [hunk 4]
```diff
@@ -43,8 +44,11 @@ type OptionalConfigParams struct {
 }
 
 var (
-	errCameraMustHaveName    = errors.New("\"camera[name]\" is required")
-	errSensorsMustNotBeEmpty = errors.New("\"sensors\" must not be empty")
+	errCameraMustHaveName              = errors.New("\"camera[name]\" is required")
+	errSensorsMustNotBeEmpty           = errors.New("\"sensors\" must not be empty")
+	errLocalizationInOfflineModeSuffix = "= 0, enable_mapping = false, and existing_map provided. localization in offline mode not supported."
```
Comment (EshaMaharishi) [inline config/config.go:49]: should this message be updated to not mention existing_map?

File: config/config.go [hunk 5]
```diff
@@ -43,8 +44,11 @@ type OptionalConfigParams struct {
...
+	errLocalizationInOfflineMode       = newError(fmt.Sprintf("data_rate_msec %s", errLocalizationInOfflineModeSuffix))
+	errLocalizationInOfflineModeIMU    = newError(fmt.Sprintf("camera[data_freq_hz] %s", errLocalizationInOfflineModeSuffix))
```
Comment (EshaMaharishi) [inline config/config.go:51]: [optional] if we do need two messages, suggestion to just write the full message out twice for easier readability

File: config/config.go [hunk 6]
```diff
@@ -179,3 +188,18 @@ func GetOptionalParameters(config *Config, defaultLidarDataRateMsec, defaultIMUD
...
+	optionalConfigParams OptionalConfigParams,
+	imuEnabled bool,
+) error {
+	offlineMode := optionalConfigParams.LidarDataRateMsec == 0
+	localizationMode := !optionalConfigParams.EnableMapping
+	if localizationMode && offlineMode {
+		if imuEnabled {
+			return errLocalizationInOfflineModeIMU
+		}
+		return errLocalizationInOfflineMode
```
Comment (EshaMaharishi) [inline config/config.go:202]: i'm also confused why we return errLocalizationInOfflineMode that mentions data_rate_msec in its message when we aren't checking data_rate_msec at all?

File: config/config_test.go [hunk 7]
```diff
@@ -297,6 +297,37 @@ func getOptionalParametersTestHelper(
 		}
 	})
 
+	t.Run(fmt.Sprintf("config that puts cartographer in offline mode and in localization mode %s", suffix), func(t *testing.T) {
+		cfgService := makeCfgService(imuIntegrationEnabled, cloudStoryEnabled)
+		cfgService.Attributes["existing_map"] = "test-file.pbstream"
+		cfgService.Attributes["data_rate_msec"] = 0
```
Comment (EshaMaharishi) [inline config/config_test.go:303]: why does the config have both data_rate_msec and camera.data_frequency_hz? shouldn't it have one or the other?

Comment (kim-mishra) [reply config/config_test.go:303]: `data_rate_msec` gets ignored if IMU is enabled, but I can only add it to the config if imu is enabled if you think that will make it clearer

File: config/config.go (hunk 5 above)
Comment (kim-mishra) [reply config/config.go:51]: This caused a lint error for too many characters in the line

File: config/config.go (hunk 6 above)
Comment (kim-mishra) [reply config/config.go:202]: discussed offline

File: config/config_test.go (hunk 7 above)
Comment (kim-mishra) [reply config/config_test.go]: made this change!

File: config/config.go (hunk 6 above)
Comment (EshaMaharishi) [reply config/config.go:200]: when IMU is not enabled, `optionalConfigParams.LidarDataRateMsec` comes from `data_rate_msec`

**Review by EshaMaharishi — APPROVED** [review]
Lgtm!

**Pull request merged**